packages/github-actions/teams-toolkit-cicd/src/operations.ts

```diff
@@ -0,0 +1,113 @@
+import * as path from 'path'
+import {Capability} from './enums/capabilities'
+import {Execute} from './utils/exec'
+import {Commands, Pathes, Miscs, ActionOutputs} from './constant'
+import {ProgrammingLanguage} from './enums/programmingLanguages'
+import * as fs from 'fs-extra'
+import * as core from '@actions/core'
+import {LanguageError} from './errors'
+
+// eslint-disable-next-line @typescript-eslint/no-extraneous-class
+export class Operations {
+  static async BuildTeamsApp(
+    projectRoot: string,
+    capabilities: string[]
+  ): Promise<void> {
+    const tabsPath = path.join(projectRoot, Capability.Tabs)
+    if (
+      capabilities.includes(Capability.Tabs) &&
+      (await fs.pathExists(tabsPath))
+    ) {
+      await Execute(Commands.NpmInstall, tabsPath)
+      await Execute(Commands.NpmRunBuild, tabsPath)
+    }
+
+    const botPath = path.join(projectRoot, Capability.Bot)
+    if (
+      capabilities.includes(Capability.Bot) &&
+      (await fs.pathExists(botPath))
+    ) {
+      // Get bot's programming language from env.default.json.
+      const config = await fs.readJSON(Pathes.EnvDefaultJson)
+      const lang = config?.[Miscs.BotConfigKey]?.[Miscs.LanguageKey]
+      if (!lang || !Object.values<string>(ProgrammingLanguage).includes(lang)) {
```

Comment:
better to check the condition before npm ops of `tabs`. And we don't need to await for tab's npm ops. The logic should be refined for better concurrency.

Reply:
It's a good idea to improve latency by running commands in parallel. Let me check it.